### PR TITLE
dvc: introduce merge-driver

### PR DESCRIPTION
**Pull request by efiop**
Related to #4162

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

TODO:
- [ ] more tests

### CODE REVIEW

File: dvc/dvcfile.py [hunk 1]
```diff
@@ -239,6 +259,9 @@ def remove_stage(self, stage):
         else:
             super().remove()
 
+    def merge(self, ancestor, other):
+        raise NotImplementedError
+
```
Comment (efiop) [inline dvc/dvcfile.py:264]: These stubs are for the future, as merging dvcfiles is potentially a handy thing to have.

File: dvc/cache/base.py [hunk 2]
```diff
@@ -550,3 +555,70 @@ def get_files_number(self, path_info, hash_, filter_info):
...
+        result = [
+            {
+                self.tree.PARAM_RELPATH: relpath,
+                self.tree.PARAM_CHECKSUM: checksum,
+            }
+            for relpath, checksum in merged.items()
+        ]
+
+        # Sorting the list by path to ensure reproducibility
+        return sorted(result, key=itemgetter(self.tree.PARAM_RELPATH))
```
Comment (efiop) [inline dvc/cache/base.py:605]: We have a similar thing in _collect_dir in BaseTree's, but these will be unified as a part of https://github.com/iterative/dvc/issues/4144 , once dir hash computation is detached from cache.

**Comment by pared** [comment]
@efiop 
What about conflicting files? I prepared sample script:
```
#!/bin/bash

rm -rf repo
mkdir repo

pushd repo

git init --quiet
dvc init --quiet

git add -A
git commit -am "init"

dvc install
echo "*.dvc merge=dvc" >> .gitattributes
git add -A
git commit -m "setup hooks"

git checkout -b one
mkdir data
echo foo >> data/foo

dvc add data
git add -A
git commit -m "add foo"

git checkout master

git checkout -b two
mkdir data
echo bar >> data/foo

dvc add data
git add -A
git commit -m "add foo but bar"

git checkout master

git merge one --no-edit
git merge two --no-edit
```
Where I have 2 branches:
`one` that has `dvc add`-ed `data/foo` with `foo` content
`two` that has `dvc add`-ed `data/foo` with `bar` content

and then, on master I merge `one` and then `two`. 
My `data.dvc` gets silently overridden so that `data/foo` contains `bar` - the `two` branch content.
Shouldn't we get some error?

**Comment by efiop** [comment]
@pared 

>git checkout master
> git merge one --no-edit
> # I think we should get error now

Not really, because there is nothing in master, so there is no conflict.

> git merge two --no-edit

And here you get a conflict, as expected.

**Comment by pared** [comment]
@efiop sorry, I meant that I would expect error after the comment.
And yes, there is conflict, but my `.dvc` file gets Auto-merged. It is some indicator of failure, but it is easy to omit. I would expect that git will not try to auto-merge. Choosing which file is the proper one should be users responsibility. 
Now the content of `data/foo` depends on the order of `merge`:
```
git merge one --no-edit
git merge two --no-edit
# cat data/foo == bar
```

```
git merge two --no-edit
git merge one --no-edit
# cat data/foo == foo
```
If we were doing same thing in git, we would get unresolved conflict

**Comment by efiop** [comment]
@pared Ah, got it. Yes, that's a bug, thank you! :pray: Finally getting to creating additional tests for it, will be sure to add one for that as well.

**Review by pared — APPROVED** [review]

**Pull request merged**